designs/workflow-refactor.md

```diff
@@ -0,0 +1,296 @@
+* Refactoring of parsing, defaulting and validating input specific to providers.
+* Refactoring of provider interfaces to align with workflow expectations (will require adjustment of provider implementations to satisfy the interfaces).
+* Documenting workflow tasks using native go docs so that provider developers have a clear understanding of the context in which they are executed.
+
+#### Out of Scope
+
+* Holistic refactoring of the cluster management component. The cluster management component has a hard to pin down scope of concerns that are heavily leveraged by workflows and it would take a broader effort to break them apart and organize that likely isn’t required for these proposed changes.
+* A plug-able provider interface that can be used to develop off-tree providers. Off-tree provider development requires much broader thought such as how our API should be structured to enable API extension, how to version the APIs and ensure they function together, and how to build and distribute artifacts specific to a given provider.
+* Refactoring the mechanisms for workflow checkpoint-recovery. The interfaces for checkpoint-recovery interaction with workflows may require adjusting but the fundamentals of the design should not change.
+
+## Proposal
+
+The following sections describe smaller proposals that aim to better define concerns for workflows and providers, provide mechanisms for hooking into or extending workflows, and reduce cognitive load when understanding workflow execution. Some sections work together with others and some are more independent.
+
+These proposals also invokes a fundamental change in how we think about providers. Specifically, providers have historically been modeled as data providers for centralized orchestration logic. The data provider approach is built on assumptions around how providers are composed with a desire to treat everything the same; this has since been proven false. This proposal seeks to change providers to take on greater functional responsibility such as installation of provider specific components to an environment. This sets into motion an expectation that providers will handle their specifics with an orchestration layer telling them when to run.
+
+### Separation of Management and Workload workflows
+
+We propose separating management and workload cluster workflows and constraining tasks to only those required for cluster management operations. Additional tasks supported by EKS-A will be moved to an extensible post-workflow hook that we can continually build on to add more features separate from the minimal cluster provisioning tasks.
+
+Having implemented this change, there is an expectation that workflow maintainers follow a documented set of guidelines on the granularity of tasks and what tasks are appropriate for featuring explicitly in a workflow.
+
+#### Management cluster workflows
+
+Management cluster workflow will comprise of the following tasks.
```

Comment:
our workflows also include error handling and responses in the form of a '[diganostics](https://github.com/aws/eks-anywhere/blob/main/pkg/workflows/diagnostics.go)' workflow task; curious how you're thinking about workflow execution error handling and diagnostic collection in this new format

Reply:
just saw https://github.com/aws/eks-anywhere/pull/2869/files#diff-82587cc580756a9e10951ee16accff7bebf69d3119e5ce253579df0015d636c2R266